hub/core/chunk_engine.py

```diff
@@ -0,0 +1,411 @@
+from hub.core.compression import decompress_array
+from math import ceil
+from typing import List, Optional, Sequence, Union, Tuple
+from hub.util.exceptions import (
+    CorruptedMetaError,
+    DynamicTensorNumpyError,
+)
+from hub.core.meta.tensor_meta import TensorMeta
+from hub.core.index.index import Index
+from hub.util.keys import (
+    get_chunk_key,
+    get_chunk_id_encoder_key,
+    get_tensor_meta_key,
+)
+from hub.core.sample import Sample  # type: ignore
+from hub.constants import DEFAULT_MAX_CHUNK_SIZE, UNCOMPRESSED
+
+import numpy as np
+
+from hub.core.storage.lru_cache import LRUCache
+
+from hub.core.chunk import Chunk
+
+from hub.core.meta.encode.chunk_id import ChunkIdEncoder
+
+
+SampleValue = Union[np.ndarray, int, float, bool, Sample]
+
+
+def is_uniform_sequence(samples):
+    """Determines if a sequence of samples has uniform type and shape, allowing it to be vectorized by `ChunkEngine.extend`."""
+    if len(set(map(type, samples))) != 1:
+        # Cannot vectorize sequence with inconsistent types
+        return False
+    elif any(isinstance(s, np.ndarray) for s in samples):
+        # Numpy arrays will only be vectorized if they have the same shape
+        return len(set(s.shape for s in samples)) == 1
+    elif any(isinstance(s, Sample) for s in samples):
+        # Sample objects will not be vectorized
+        return False
+    else:
+        # Scalar samples can be vectorized
+        return True
+
+
+class ChunkEngine:
```

Comment:
a lot of methods in this seem long and kinda hard to read

Reply:
we can address this further in the refactor sprint. unless you diagree, i think it's legible enough for a tentative approval